diff --git a/mfd-ch347.c b/mfd-ch347.c
index f7057b3..a9d4ad8 100644
--- a/mfd-ch347.c
+++ b/mfd-ch347.c
@@ -34,6 +34,7 @@
 
 #define CH347_USB_VENDOR 0x1a86
 #define CH347_USB_DEVICE_M1 0x55db
 #define CH347_USB_DEVICE_M3 0x55dd
+#define CH347_USB_DEVICE_M4 0x55de
 
 #define CH347_DEFAULT_TIMEOUT 1000
@@ -725,6 +726,7 @@ static int ch347_probe(struct usb_interface *interface, const struct usb_device
 static const struct usb_device_id ch347_table[] = {
 	{ USB_DEVICE(CH347_USB_VENDOR, CH347_USB_DEVICE_M1) }, /* mode #1 (UART + SPI + I2C + GPIO) */
 	{ USB_DEVICE(CH347_USB_VENDOR, CH347_USB_DEVICE_M3) }, /* mode #3 (UART + JTAG + I2C + GPIO) */
+	{ USB_DEVICE(CH347_USB_VENDOR, CH347_USB_DEVICE_M4) }, /* mode #1, newer firmware (UART + SPI + I2C + GPIO) */
 	{ }
 };
